projects/plugins/protect/src/js/hooks/use-onboarding/index.jsx

```diff
@@ -0,0 +1,100 @@
+import { useDispatch, useSelect } from '@wordpress/data';
+import { createContext, useCallback, useContext, useEffect, useMemo, useState } from 'react';
+import API from '../../api';
+import { STORE_ID } from '../../state/store';
+import useProtectData from '../use-protect-data';
+
+export const OnboardingContext = createContext( [] );
+export const OnboardingRenderedContext = createContext( [] );
+
+export const OnboardingRenderedContextProvider = ( { children } ) => {
+	const [ renderedSteps, setRenderedSteps ] = useState( [] );
+
+	return (
+		<OnboardingRenderedContext.Provider value={ { renderedSteps, setRenderedSteps } }>
+			{ children }
+		</OnboardingRenderedContext.Provider>
+	);
+};
+
+const useOnboarding = () => {
+	const { hasRequiredPlan } = useProtectData();
+	const { completeOnboardingSteps, fetchOnboardingProgress } = API;
+
+	const steps = useContext( OnboardingContext );
+	const { renderedSteps } = useContext( OnboardingRenderedContext );
+
+	const progress = useSelect( select => select( STORE_ID ).getOnboardingProgress() );
+	const { setOnboardingProgress } = useDispatch( STORE_ID );
+
+	/**
+	 * Current Step
+	 *
+	 * @returns {null|object}
+	 */
+	const { currentStep, currentStepCount, stepsCount } = useMemo( () => {
+		return steps.reduce(
+			( carry, step ) => {
+				if ( renderedSteps.includes( step.id ) ) {
+					carry.stepsCount++;
+					if ( ! carry.currentStep && ( progress || [] ).indexOf( step.id ) === -1 ) {
+						carry.currentStep = step;
+						carry.currentStepCount = carry.stepsCount;
+					}
+				}
+				return carry;
+			},
+			{
+				currentStep: null,
+				currentStepCount: null,
+				stepsCount: 0,
+			}
+		);
+	}, [ progress, renderedSteps, steps ] );
+
+	const completeCurrentStep = useCallback( () => {
+		if ( currentStep ) {
+			// Complete the step immediately in the UI
+			setOnboardingProgress( [ ...progress, currentStep.id ] );
+			// Save the completion in the background
+			completeOnboardingSteps( [ currentStep.id ] );
+		}
+	}, [ currentStep, setOnboardingProgress, progress, completeOnboardingSteps ] );
+
+	/**
+	 * Complete All Current Steps
+	 */
+	const completeAllCurrentSteps = useCallback( () => {
+		const stepIds = steps
+			.filter( step => {
+				// If hasRequiredPlan is true, include paid steps, else include free steps
+				return hasRequiredPlan ? step.id.startsWith( 'paid-' ) : step.id.startsWith( 'free-' );
+			} )
+			.reduce( ( carry, step ) => {
+				carry.push( step.id );
+				return carry;
+			}, [] );
```

Comment:
```suggestion
			.map( step => step.id );
```
Nitpick - good use case for `map`.

Reply:
@nateweller I am realizing now that we could make this even more efficient and only iterate over the array a single time with something along the lines of:
```
const stepIds = steps.reduce( ( carry, step ) => {
	if ( hasRequiredPlan ? step.id.startsWith( 'paid-' ) : step.id.startsWith( 'free-' ) ) {
		carry.push( step.id );
	}
	return carry;
}, [] );
```
Does that appeal to you?